dbms/src/Storages/MergeTree/MergeTreeBaseBlockInputStream.cpp

```diff
@@ -131,7 +132,18 @@ Block MergeTreeBaseBlockInputStream::readFromPart()
             {
                 auto & col = block.getByPosition(i);
                 if (task.column_name_set.count(col.name))
-                    col.column = col.column->cloneEmpty();
+                {
+                    if (ColumnArray * column_array = typeid_cast<ColumnArray *>(col.column.get()))
+                    {
+                        /// ColumnArray columns in block have common offset column, which is used while reading.
```

Comment:
This is true only for parts of "nested data structures".
Ordinary Arrays could have different sizes, obviously.